### DISCUSSION TITLE
Face annotations in file metadata, seeking advice and standards

### DISCUSSION THREAD

**Discussion by Xavier-LB** [annotorious/annotorious#327] [Q&A] [answered]
Hi, I'm working on an app ([OneFolder](https://onefolder.app)) for managing photos (similar to [Google Photos](https://photos.google.com) or [Apple Photos](https://www.icloud.com/photos)) but without a server, relying solely on file metadata. The app uses [ExifTool](https://exiftool.org/), adding some tags to the picture metadata, such as `Subject` and `Keywords`.

Now, I want to annotate faces in the file's metadata using [Annotorious](https://github.com/annotorious/annotorious).

What I need to know now is whether I should try to adhere to this [W3C WebAnnotation format](https://www.w3.org/TR/annotation-model/) when adding it to the file's metadata. If yes, does [ExifTool](https://exiftool.org/#links) provide features to work with or support the [W3C WebAnnotation format](https://www.w3.org/TR/annotation-model/) in any way?

If no, which metadata standard should I use?

I tried to find examples from other similar apps like [DigiKam](https://www.digikam.org/), [PiGallery2](https://github.com/bpatrik/pigallery2), and [Photheca](https://lunarship.com/). I assumed that, in most cases, apps use [XMP](https://www.adobe.com/products/xmp.html) respecting [Metadata Working Group (MWG) format](https://onefolder.app/mwg.pdf), even if it appears that they are no longer active (their website http://www.metadataworkinggroup.org/ does not work).

My real intention is to respect the metadata standards for photos as much as possible. I don't want to create more confusion than there already is.

Is there an existing app that uses [Annotorious](https://github.com/annotorious/annotorious), annotations and **add it** in the metadata?

**Comment by rsimon** [comment, marked as answer]
Hi, this looks fantastic! ❤️ 

I remember having a chat with someone about encoding regions in EXIF tags. That was years ago though... I think XMP was also what they mentioned. (Probably [this part about XMP region tags](https://exiftool.org/TagNames/MWG.html#Regions), which looks like it's the same of what you have in mind.) 

I don't think Web Annotation would be suitable. It's quite specialized for online content after all, and leans on standards that make sense in the browser (such as identifying everything through URIs, encoding rectangles as [media fragments](https://www.w3.org/TR/media-frags/#naming-space), etc.). But in the context of EXIF, XMP feels like the way to go.

I'm not aware of any app that currently uses Annotorious for EXIF tagging. You'd definitely need to build a bit of crosswalking between formats. Shouldn't be too difficult, however, since media fragments are really just concatenated x/y/width/height tuples.

BTW: if you are building your app in React (which I think you do) - the [next major release of Annotorious](https://github.com/annotorious/annotorious/tree/3.x) is almost ready. It's still in alpha state, and currently undocumented... nonetheless, it's a lot easier to use with frameworks, especially if you're not planning to use the default Annotorious comment popup. (There's a dedicated [@annotorious/react package](https://www.npmjs.com/package/@annotorious/react).) Let me know if you would like to give that a try, and I can help out.

**Reply by Xavier-LB** [reply]
Thanks, I had my doubts about trying to use W3C or not. But I think for this case the best is to add the information I need with MWG Regions Tags.
I will try to make a blog to explain how I use Annotorious to add faces in the metadata, I think it can be quite useful.

We hesitated to use v3 because it looked to early, but if you tell us that we should go with that, and if we can ask you some question if we are blocked, we would happy to jump into v3 right away

**Reply by rsimon** [reply]
A blog post would be fantastic - thanks!

v3 is more risky for sure at this point. Due to lack of documentation, you would have to dig into [the source code](https://github.com/annotorious/annotorious/tree/3.x/packages/annotorious-react/src). And I can't really offer much of a code example except [this very simple one](https://github.com/rsimon/immarkus/blob/13b30fb9717ae690af2834632552dcc9a6c9f5c0/src/pages/annotate/Annotate.tsx#L69-L78). I'm definitely available to ask if you run into problems though. The advantage, I think, would be that you wouldn't have to create your own React wrapper. (And the auto-save feature is probably nice, too. It will trigger annotation create/update events as soon as the user stops moving/resizing the annotation. No need for an extra click on a Save button, an explicit 'de-select' action or anything.)

**Reply by rsimon** [reply]
P.S.: I just put the minimal possible `@annotorious/react` example here: 
https://codesandbox.io/p/sandbox/fervent-thunder-xkpxvc
